server/midddleware/forwarder.go

```diff
@@ -86,9 +123,98 @@ func forwarderUnaryServerInterceptor() grpc.UnaryServerInterceptor {
 		txCtx := api.GetTransaction(ctx, req.(proto.Message))
 
 		if txCtx != nil && txCtx.GetOrigin() != types.MyOrigin {
-			return forwardRequest(ctx, txCtx, info.FullMethod, req.(proto.Message))
+			iface, err := forwardRequest(ctx, txCtx, info.FullMethod, req.(proto.Message))
+			log.Err(err).Str("method", info.FullMethod).Msg("forwarded request")
+			return iface, err
 		}
 
 		return handler(ctx, req)
 	}
 }
+
+func proxyHandler(_ interface{}, serverStream grpc.ServerStream) error {
+	fullMethodName, ok := grpc.MethodFromServerStream(serverStream)
+	if !ok {
+		return status.Errorf(codes.Internal, "lowLevelServerStream not exists in context")
+	}
+
+	outgoingCtx, backendConn, req, resp, err := proxyDirector(serverStream.Context(), fullMethodName)
+	if err != nil {
+		return err
+	}
+
+	clientCtx, clientCancel := context.WithCancel(outgoingCtx)
+	defer clientCancel()
+
+	clientStream, err := grpc.NewClientStream(clientCtx, &grpc.StreamDesc{ServerStreams: true, ClientStreams: true}, backendConn, fullMethodName)
+	if err != nil {
+		return err
+	}
+
+	s2cErrChan := forwardServerToClient(serverStream, req, clientStream)
+	c2sErrChan := forwardClientToServer(clientStream, resp, serverStream)
+
+	for i := 0; i < 2; i++ {
+		select {
+		case s2cErr := <-s2cErrChan:
+			if s2cErr == io.EOF {
+				_ = clientStream.CloseSend()
+			} else {
+				clientCancel()
+				return status.Errorf(codes.Internal, "failed proxying s2c: %v", s2cErr)
+			}
+		case c2sErr := <-c2sErrChan:
+			serverStream.SetTrailer(clientStream.Trailer())
+			if c2sErr != io.EOF {
+				return c2sErr
+			}
+			return nil
+		}
+	}
+	return status.Errorf(codes.Internal, "gRPC proxying should never reach this stage.")
+}
+
+func forwardClientToServer(src grpc.ClientStream, resp proto.Message, dst grpc.ServerStream) chan error {
+	ret := make(chan error, 1)
+	go func() {
+		for i := 0; ; i++ {
+			if err := src.RecvMsg(resp); err != nil {
+				ret <- err
+				break
+			}
+			if i == 0 {
+				md, err := src.Header()
+				if err != nil {
+					ret <- err
+					break
+				}
+				if err := dst.SendHeader(md); err != nil {
+					ret <- err
+					break
+				}
+			}
+			if err := dst.SendMsg(resp); err != nil {
+				ret <- err
+				break
+			}
+		}
+	}()
+	return ret
+}
+
+func forwardServerToClient(src grpc.ServerStream, req proto.Message, dst grpc.ClientStream) chan error {
+	ret := make(chan error, 1)
+	go func() {
+		for i := 0; ; i++ {
```

Comment:
Why not just for {}

Reply:
I is used to send headers on first message.

Reply:
where is that?

Reply:
ah, its in other function. removing this one.